### ISSUE TITLE
Question about pruning ratio implementation

### ISSUE THREAD

**Issue by lky-violet** [HVision-NKU/GlimpsePrune#12]
Hello, thank you for your excellent work!  

In the *Visual importance prediction and one-shot pruning* section, the paper mentions that *“maximum retention ratio can be configured to sparsify the visual tokens”*.  

When checking the implementation, I found the following code:  

```python
one_image_token_bool_indices = torch.topk(one_image_token_prob, max_remain_num, dim=-1).indices
one_image_token_bool_mask.zero_()
one_image_token_bool_mask[one_image_token_bool_indices] = True
```

I noticed that in ATP-llava and dynamic-llava, using a hard threshold or argmax can lead to non-differentiable operations, which block gradient backpropagation.

Did you notice this issue in your implementation as well?  

Thank you again for your patience!

**Comment by lky-violet** [comment]
Thank you very much for your reply!   

However, I still have some doubts.  How is the pruning rate controlled during training? If we directly select based on a threshold, wouldn’t there still be an issue with gradient propagation?

```python
def _get_remain_masks(self, input_ids, attention_mask, image_token_mask_logits, grid_hw):
        threshold = self.config.reduce_threshold
        anchor_positions = list(self.config.anchor_positions) if self.config.anchor_positions is not None else []
        min_remain_num = self.config.min_remain_num
        max_remain_ratio = self.config.max_remain_ratio
        
        image_token_bool_masks = []

        last_image_token_mask_logits = image_token_mask_logits[:, -1]  # B,N

        for b, one_logits in enumerate(last_image_token_mask_logits):
            one_image_token_prob = one_logits.sigmoid()  # [N,]
            one_image_token_bool_mask = one_image_token_prob > threshold
            
            if max_remain_ratio is not None:
                remain_num = one_image_token_bool_mask.sum().item()
                remain_ratio = remain_num / one_image_token_bool_mask.numel()
                if remain_ratio > max_remain_ratio:
                    max_remain_num = int(max_remain_ratio * one_image_token_bool_mask.numel())
                    one_image_token_bool_indices = torch.topk(one_image_token_prob, max_remain_num, dim=-1).indices
                    one_image_token_bool_mask.zero_()
                    one_image_token_bool_mask[one_image_token_bool_indices] = True
            
            if min_remain_num is not None:
                remain_num = one_image_token_bool_mask.sum().item()
                if remain_num < min_remain_num:
                    one_image_token_bool_indices = torch.topk(one_image_token_prob, min_remain_num, dim=-1).indices
                    one_image_token_bool_mask[one_image_token_bool_indices] = True
            
            if anchor_positions:
                one_attn_height, one_attn_width = grid_hw
                for anchor_pos in anchor_positions:
                    if anchor_pos == 'tl':
                        one_image_token_bool_mask[0] = True
                    elif anchor_pos == 'tr':
                        anchor_pos_id = one_attn_width - 1
                        one_image_token_bool_mask[anchor_pos_id] = True
                    elif anchor_pos == 'bl':
                        anchor_pos_id = (one_attn_height - 1) * one_attn_width
                        one_image_token_bool_mask[anchor_pos_id] = True
                    elif anchor_pos == 'br':
                        anchor_pos_id = one_attn_height * one_attn_width - 1
                        one_image_token_bool_mask[anchor_pos_id] = True
                    else:
                        raise ValueError(f"Unknown anchor position: {anchor_pos}. Supported: tl, tr, bl, br.")
            
            image_token_bool_masks.append(one_image_token_bool_mask)
            
        is_image_mask = input_ids == IMAGE_TOKEN_INDEX
        remain_masks = attention_mask.clone().bool()
        remain_masks[is_image_mask] = torch.cat(image_token_bool_masks, dim=0)
        remain_masks &= attention_mask.bool()  # Ensure remain_masks is still a valid attention mask, usually all image tokens are valid in attention_mask
        return remain_masks, image_token_bool_masks
```

**Comment by ashun989** [comment]
Below is my response to the question of how gradients are propagated during training:

1. The gradient of language loss ($\mathcal{L}_\text{lang}$) acts directly on the glimpse token. However, since the process of binarization prediction mask is not differentiable (that is, the code you cut out here), the language loss cannot optimize the parameters of the VIP part.
2. The gradient of location loss ($\mathcal{L}_\text{loc}$) can directly optimize the `image_token_mask_logits` output by the VIP, and thereby also indirectly optimize the glimpse token.
3. The combined effect of the two optimization goals enables the output of VIP to dynamically cover the important tokens. The purpose of setting `max_remain_ratio` is mainly to facilitate comparison with the fixed compression rate method.
